File: 503. Next Greater Element II/Python/Solution.py
class Solution(object):
    def nextGreaterElements(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        stack = []
        for i in range(len(nums) - 1, -1, -1):
            stack.append(i)
        result = [None] * len(nums)
        for i in range(len(nums) - 1, -1, -1):
            while stack and nums[stack[-1]] <= nums[i]:
                stack.pop()
            if stack:
                result[i] = nums[stack[-1]]
            else:
                result[i] = -1
            stack.append(i)
        return result

File: 503. Next Greater Element II/C++/Solution.cpp
class Solution {
public:
    vector<int> nextGreaterElements(vector<int> &nums) {
        stack<int> sta;
        for (int i = nums.size() - 1; i >= 0; i--) {
            sta.push(i);
        }
        vector<int> result(nums.size());
        for (int i = nums.size() - 1; i >= 0; i--) {
            while ((!sta.empty()) && nums[sta.top()] <= nums[i]) {
                sta.pop();
            }
            if (!sta.empty()) {
                result[i] = nums[sta.top()];
            } else {
                result[i] = -1;
            }
            sta.push(i);
        }
        return result;
    }
};